### ISSUE TITLE
OSM keys are split on ":", which should be treated as a literal

### ISSUE THREAD

**Issue by axb21** [gama-platform/gama#3102]
I am posting this issue on the suggestion of @AlexisDrogoul, following a discussion in the Google group that can be consulted here: https://groups.google.com/g/gama-platform/c/l5jzWUYgL8s

**Describe the bug**
Certain OSM entities like power stations have literal keys such as "generator:source" containing colons, or keys that represent "subtype" such as "name:de" (the name of the entity in German). GAMA exhibits undesirable behavior on such keys, at least in the cases described here.

**To Reproduce**
1. Create `FileTest.gaml` from the listing in the **Additional context** section below
2. Create file `tmi-example.osm` from the listing below and place it in `../includes/` relative to `FileTest.gaml`.
3. Click "parse_file" from the `FileTest.gaml` tab
4. Observe this output in the console:
```
name: Three Mile Island jednotka 1
operator: Exelon Nuclear
source: null
```

**Expected behavior**
The expect output from the above procedure is:
```
name: Three Mile Island Unit 1
operator: Exelon Nuclear
source: nuclear
```
since the GAML code explicitly requested the value of the "name" key, not the "name:sk" key; and it requested the value of the "generator:source" key, which is "nuclear".

**Desktop (please complete the following information):**
 - OS: Kubuntu 20.04
 - GAMA version: 1.8.1
 - Java version: verified issue on 1.8 and 1.11

**Additional context**
`FileTest.gaml`
```
model FileTest

global {
	init {
		file<geometry> osmfile <- file<geometry>(osm_file("../includes/tmi-example.osm"));
		create tmi_agent from: osmfile with: [name:: read("name"), operator::read("operator"), generator_source::read("generator:source")];
		loop the_agent over: tmi_agent {
			write ("name: " + the_agent get "name");
			write ("operator: " + the_agent get "operator");
			write ("source: " + the_agent get "generator_source");
			break;
		}
	}
}

species tmi_agent {
	string name;
	string operator;
	string generator_source;
}

experiment "parse_file" {}
```

`tmi-example.osm`:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<osm version="0.6" generator="CGImap 0.8.3 (1663153 spike-07.openstreetmap.org)" copyright="OpenStreetMap and contributors" attribution="http://www.openstreetmap.org/copyright" license="http://opendatacommons.org/licenses/odbl/1-0/">
 <bounds minlat="40.1459000" minlon="-76.7293000" maxlat="40.1603000" maxlon="-76.7197000"/>
 <node id="33180844" visible="true" version="9" changeset="92771479" timestamp="2020-10-20T13:15:59Z" user="Olyon" uid="1443767" lat="40.1547531" lon="-76.7251031">
  <tag k="disused:power" v="generator"/>
  <tag k="end_date" v="2019-09-20"/>
  <tag k="generator:method" v="fission"/>
  <tag k="generator:output:electricity" v="802 MW"/>
  <tag k="generator:source" v="nuclear"/>
  <tag k="name" v="Three Mile Island Unit 1"/>
  <tag k="name:de" v="Three Mile Island Block 1"/>
  <tag k="name:sk" v="Three Mile Island jednotka 1"/>
  <tag k="operator" v="Exelon Nuclear"/>
  <tag k="short_name" v="TMI-1"/>
  <tag k="website" v="http://www.exeloncorp.com/PowerPlants/threemileisland/"/>
 </node>
</osm>
```

@AlexisDrogoul pointed out in the Google group that [GamaOsmFile.java](https://github.com/gama-platform/gama/blob/master/msi.gama.core/src/msi/gama/util/file/GamaOsmFile.java) is splitting keys on colons, which is likely to be the source of the issue described above. You can see this in 4 places: on line 442, in the parsing o Node tags; on line 498, in the parsing of Way tags; on line 594 , in the parsing of Relations; and again on line 662. Unless I'm misremembering, in all places the split code looks something like this:

```java
final String key = tg.getKey().split(":")[0];
```

where `tg` holds a Tag line such as:

```xml
<tag k="name:sk" v="Three Mile Island jednotka 1"/>
```

from the example. This means that any key of form "foo:bar:baz..." will be split on colons, resulting in an array ["foo","bar","baz",...], and then reduced to "foo" by [0], which accesses the first element. "name", "name:de", "name:sk" will therefore all be treated as having key "name" and, since "name:sk" is last and the file is processed element-by-element, "name:sk" will become the value for `read("name")`. Whatever the intended behavior of splitting by ":" was, this conflation of distinct keys cannot be correct.

I am unfamiliar with this codebase, but from my naive perspective it seems the four offending lines should read something like

```java
final String key = tg.getKey();
```

instead, where the `split(":")[0]` has been removed. That change would likely fix the issue I describe above, but of course could have a deleterious effect on use cases I'm unfamiliar with.

**Comment by axb21** [comment]
I neglected to mention that as a workaround, `.osm` files can be preprocessed to change ":" to some other symbol. I have verified that "-" works. Of course, the corresponding `.gaml` code will have to be changed accordingly. This workaround does not help with `.osm.pbf` files, which are binary, or with compressed `.osm` files unless they are uncompressed, preprocessed, and the re-compressed.

**Comment by AlexisDrogoul** [comment]
Update (sorry for the long delay): have you proposed your pull request ? @ptaillandier what's your opinion ?

**Comment by axb21** [comment]
> Update (sorry for the long delay): have you proposed your pull request ? @ptaillandier what's your opinion ?

Hi @AlexisDrogoul, I just created the pull request. I had hesitated to do that because I was unsure whether you'd want that. 

For what it's worth, I have verified this code compiles with jdk 1.8 on Ubuntu 20.10 (using `./travis/compile.sh`). I have also verified it does not cause any new test projects to fail. I'm happy to provide additional help if needed.
